src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs

```diff
@@ -491,9 +511,28 @@ override public void Close()
                     ADP.TimerCurrent(out _statistics._closeTimestamp);
                 }
             }
+            catch (Exception ex)
+            {
+                e = ex;
+                throw;
+            }
             finally
             {
                 SqlStatistics.StopTimer(statistics);
+
+                // we only want to log this if the previous state of the 
+                // connection is open, as that's the valid use-case
+                if (previousState == ConnectionState.Open)
+                { 
+                    if (e != null)
+                    {
+                        s_diagnosticListener.WriteConnectionCloseError(operationId, clientConnectionId, this, e);
+                    }
+                    else
+                    {
+                        s_diagnosticListener.WriteConnectionCloseAfter(operationId, clientConnectionId, this);
```

Comment:
FWIW, I still prefer the pattern of just exposing an after method rather than both an after and error method, but i realize others disagree.